src/main/java/org/launchcode/recipeapp/controllers/SearchController.java

```diff
@@ -20,20 +16,277 @@
+                model.addAttribute("sort", SortParameter.values());
+
+                // filter recipes from CATEGORY search OR KEYWORD search
+            } else if (!foundRecipes.isEmpty()) {
+                sortedRecipes.clear();
+                Iterable<Recipe> recipes = foundRecipes;
+                for (Recipe recipe : recipes) {
+                    if (recipe.getTag().getName().toLowerCase().equals(filterValue.getName().toLowerCase())) {
+                        filteredRecipes.add(recipe);
+
+                        //render filtered recipes from category search OR keyword search
+                        model.addAttribute("recipes", filteredRecipes);
+                        model.addAttribute("categories", Category.values());
+                        model.addAttribute("sort", SortParameter.values());
+                    }
+                }
+            } else if (!sortedRecipes.isEmpty()) {
+                //filter recipes from "SORT BY"
+                Iterable<Recipe> recipes = sortedRecipes;
+                List<Recipe> filteredRecipes = new ArrayList<>();
+                for (Recipe recipe : recipes) {
+                    if (recipe.getTag().getName().toLowerCase().equals(filterValue.getName().toLowerCase())) {
+                        filteredRecipes.add(recipe);
+
+                        //render filtered recipes from "SORT BY"
+                        model.addAttribute("recipes", filteredRecipes);
+                        model.addAttribute("categories", Category.values());
+                        model.addAttribute("sort", SortParameter.values());
+                    }
+                }
+            } else {
+                //
+                if (filteredRecipes.isEmpty()) {
+                Iterable<Recipe> recipes = recipeRepository.findAll();
+                List<Recipe> filteredRecipes = new ArrayList<>();
+                    for (Recipe recipe : recipes) {
+                    if (recipe.getTag().getName().toLowerCase().equals(filterValue.getName().toLowerCase())) {
+                        filteredRecipes.add(recipe);
+
+                        //render filtered
+                        model.addAttribute("recipes", filteredRecipes);
+                        model.addAttribute("categories", Category.values());
+                        model.addAttribute("sort", SortParameter.values());
+                         }
+                    }
+                }
+            }
+        }
+        return "search";
+    }
+
+
+    //SORT SEARCH RESULTS
+    @PostMapping(value = "/sort")
+    public String sortSearchResults(@RequestParam SortParameter sortParameter, Model model) {
+      sortedRecipes.clear();
+        //sort recipes from CATEGORY search OR KEYWORD search by ASCENDING NAME
+        if (filteredRecipes.isEmpty()) {
+            Iterable<Recipe> recipes = foundRecipes;
+            if ((sortParameter.getName().equals(SortParameter.NAME_ASCENDING.getName()))) {
```

Comment:
This can produce NullPointerException , you can be rewrite like this SortParameter.NAME_ASCENDING.getName().equals(sortParameter.getName())